src/codecs.rs

```diff
@@ -0,0 +1,120 @@
+use bytes::{BufMut, BytesMut};
+use codec::{Encoder, Decoder};
+use std::{io, str};
+
+/// A simple `Codec` implementation that just ships bytes around.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct Bytes;
+
+/// A simple `Codec` implementation that just ships bytes around.
+// Note: while this constant is technically reduntant above,
+// it's nice to include so that `codec::BYTES` and `codec::LINES`
+// provide the same interface to the end-user.
+pub const BYTES: Bytes = Bytes;
+
+impl Decoder for Bytes {
+    type Item = BytesMut;
+    type Error = io::Error;
+
+    fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<BytesMut>, io::Error> {
+        if buf.len() > 0 {
+            let len = buf.len();
+            Ok(Some(buf.split_to(len)))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn decode_eof(&mut self, buf: &mut BytesMut) -> Result<Option<BytesMut>, io::Error> {
+        self.decode(buf)
+    }
+}
+
+impl Encoder for Bytes {
+    type Item = Vec<u8>;
```

Comment:
This brings up an interesting question! I'm not sure if `Vec<u8>` or `Bytes` would perhaps be more appropriate here... I might lean a bit more towards `Bytes`, but I imagine that's a little more unergnomic?

Reply:
Indeed, I don't think it would be great to add an encoder that takes `<Vec<u8>>` as that would require allocating a vec to store the bytes that gets discarded immediately.

In general, I don't see great value to have an `Encoder` for bytes since if you already have bytes, why would you not use `io::write_all` or something like that to avoid the extra overhead?

Reply:
@carllerche In the place I wanted to use it, I was trying to make a `framed` handler which performed some action every time bytes were received, rather than writing the bytes into some type which implemented `AsyncRead`. It's possible to solve this issue by making your own type that implements `AsyncRead` and implementing the handler functions appropriately, but that's a lot more boilerplate.

Reply:
Is there a different way to solve this problem?

Reply:
Also see https://github.com/hyperium/hyper/issues/1364#issuecomment-340612348

Reply:
I'm personally ok having this in `tokio-io` as I agree it's a relatively large amount of boilerplate to get the read/write halves working whereas otherwise it's just `.framed(Bytes::new())`?

Reply:
Changed to use `Bytes`.